### DISCUSSION TITLE
Is there a way to set the Window (frame) minimum height/width? (WinUI 3)

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#9092] [Q&A]
[_from discord_](https://discord.com/channels/372137812037730304/550416151172087808/989233695212847115)

The only thing I found online was this (https://github.com/microsoft/microsoft-ui-xaml/issues/2945#issuecomment-776446712) which seems very overcomplicated, and it only works on Windows.

**Comment by jeromelaban** [comment]
Indeed, at this time, the WinUI 3 APIs are very specific to windows. Fortunately, most of those APIs were available as part of WinRT and can be used for the other platforms (where Window sizing is applicable):

```csharp
ApplicationView.GetForCurrentView().SetPreferredMinSize(new Size(640, 480));
```

**Reply by ArchLeaders** [reply]
I have placed that under `InitializeComponent()` in the startup page backend.
When running the application, an exception is thrown:
```
Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
```

**Reply by jeromelaban** [reply]
Which project are you running when you get this error?

**Reply by ArchLeaders** [reply]
The Windows platform project.

**Reply by jeromelaban** [reply]
Got it. With WinAppSDK apps, `ApplicationView.GetForCurrentView().SetPreferredMinSize` is not available and you need to use the samples provided in https://github.com/microsoft/microsoft-ui-xaml/issues/2945#issuecomment-776446712. The best way at this point is to use [conditional code](https://platform.uno/docs/articles/platform-specific-csharp.html) to switch the implementation based on the compiled project. Uno targets use `SetPreferredMinSize` and WinAppSDK ones use the Win32 APIs.

**Reply by ArchLeaders** [reply]
Very well. Thank you for your time.

**Reply by ArchLeaders** [reply]
The solution in that discussion brings about two compiler errors for me.

1. _'ShellView' does not contain a definition for 'As' and no accessible extension method 'As' accepting a first argument of type 'ShellView' could be found (are you missing a using directive or an assembly reference?)_

![image](https://user-images.githubusercontent.com/80713508/175116288-51497c28-3d2e-44ae-a3ba-9fa9768f2d8d.png)

2. _The name 'PInvoke' does not exist in the current context (this error is duplicated many times throughout) the snippet._

![image](https://user-images.githubusercontent.com/80713508/175116646-119a621d-8020-466b-8674-d54895f37ceb.png)

The only difference between my code and the snippet is that the derived type is `Page` not `Window`. But I cannot derive from Window without getting another compiler error: `'ShellView': cannot derive from sealed type 'Window'`.

**Reply by jeromelaban** [reply]
You need to make that code conditional to windows, by using `#if !HAS_UNO` for instance,

**Reply by ArchLeaders** [reply]
Ah right, didn't know that was crucial to the code. My bad.
Using the `#if WINDOWS` statement leaves just one error when getting the HWND. `Member 'Page.As<ShellView.IWindowNative>()' cannot be accessed with an instance reference; qualify it with a type name instead`. (Same error when using `!HAS_UNO` as well)

**Reply by ArchLeaders** [reply]
Using the static type `Page` to satisfy the compiler results in the same memory read access error as before.

**Comment by EvarDion** [comment]
This code didn't work for me

```csharp
ApplicationView.GetForCurrentView().SetPreferredMinSize()
```

But another option which works in WINUI3 under windows is to handle the **AppWindow.Changed** event that fires when the window changes check if the new size is bellow **MinWidth** or **MinHeight** and then resize it back to the **MinHeight** or **MinWidth**.

This isn't an elegant solution as it causes the window to flicker a bit but it does prevent the user from resizing the window smaller than is desired and I use it to set the startup size for my app.


```csharp
        /// <summary>
        /// Sets the size of the Application Window when running in Windows.
        /// </summary>
        /// <param name="window"></param>o
        /// <param name="width"></param>
        /// <param name="height"></param>
        /// <returns></returns>
        public static bool Resize(Window window, int width, int height)
        {

            try
            {

#if NET7_0_WINDOWS10_0_18362
                // Use 'this' rather than 'window' as variable if this is about the current window.
                IntPtr hWnd = WinRT.Interop.WindowNative.GetWindowHandle(window);
                Microsoft.UI.WindowId windowId = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(hWnd);
                AppWindow = Microsoft.UI.Windowing.AppWindow.GetFromWindowId(windowId);
                AppWindow.Resize(new Windows.Graphics.SizeInt32 { Width = width, Height = height });
                AppWindow.Changed += AppWindow_Changed;
#else
                _ = Log.Warn($"{nameof(WindowHelper)}.{nameof(Resize)}() is not currently supported on this platform.");

#endif

                return true;
            }
            catch (Exception ex)
            {
                _ = Log.Error(ex);
            }
            return false;
        }

        private static void AppWindow_Changed(Microsoft.UI.Windowing.AppWindow sender, Microsoft.UI.Windowing.AppWindowChangedEventArgs args)
        {
            try
            {
#if NET7_0_WINDOWS10_0_18362
                if (AppWindow == null)
                    return;

                if (AppWindow.Size.Height < MinHeight && AppWindow.Size.Width < MinWidth)
                    AppWindow.Resize(new Windows.Graphics.SizeInt32 { Width = MinWidth, Height = MinHeight });
                else if (AppWindow.Size.Height < MinHeight)
                    AppWindow.Resize(new Windows.Graphics.SizeInt32 { Width = AppWindow.Size.Width, Height = MinHeight });
                else if (AppWindow.Size.Width < MinWidth)
                    AppWindow.Resize(new Windows.Graphics.SizeInt32 { Width = MinWidth, Height = AppWindow.Size.Height });
#endif

            }
            catch (Exception ex)
            {
                _ = Log.Error(ex);
            }
        }
```

**Comment by elibroftw** [comment]
Easy as pie for WinUI 3 in 2026 if you know what you are doing. Just install `WinUIEx`

```cs
public MainWindow()
{
    InitializeComponent();
  var manager = WinUIEx.WindowManager.Get(this);
  manager.PersistenceId = "MainWindow";
  manager.MinWidth = 500;
  manager.MinHeight = 550;
}
```
